<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>One function to visualise them all — vis • immunarch</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="One function to visualise them all — vis" />

<meta property="og:description" content="Output from every function in immunarch can be visualised with a
single function — vis. The vis automatically detects
the type of the data and draws a proper visualisation. For example, output
from the repOverlap function will be identified as repertoire overlap values
and respective visualisation will be chosen without any additional arguments.
See &quot;Details&quot; for the list of available visualisations." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>

  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/1_introduction.html">Introduction to `immunarch`</a>
    </li>
    <li>
      <a href="../articles/2_data.html">Working with data in `immunarch`</a>
    </li>
    <li>
      <a href="../articles/3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li>
      <a href="../articles/4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>One function to visualise them all</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunomind/immunarch/blob/master/R/vis.R'><code>R/vis.R</code></a></small>
    <div class="hidden name"><code>vis.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Output from every function in immunarch can be visualised with a
single function — <code>vis</code>. The <code>vis</code> automatically detects
the type of the data and draws a proper visualisation. For example, output
from the <code>repOverlap</code> function will be identified as repertoire overlap values
and respective visualisation will be chosen without any additional arguments.
See "Details" for the list of available visualisations.</p>
    
    </div>

    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>Pass the output from any immunarch analysis tool to <code>vis()</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>List of available visualisations for different kinds of data.</p>
<p>Basic analysis:</p>
<p>- Exploratory analysis results (from <a href='repExplore.html'>repExplore</a>) — see <a href='vis.immunr_exp_vol.html'>vis.immunr_exp_vol</a>;</p>
<p>- Clonality statistics (from <a href='repClonality.html'>repClonality</a>) — see <a href='vis.immunr_clonal_prop.html'>vis.immunr_homeo</a>.</p>
<p>Overlaps and public clonotypes:</p>
<p>- Overlaps (from <a href='repOverlap.html'>repOverlap</a>) using heatmaps, circos plots, polar area plots — see <a href='vis.immunr_ov_matrix.html'>vis.immunr_ov_matrix</a>;</p>
<p>-  Overlap clustering (from <a href='repOverlapAnalysis.html'>repOverlapAnalysis</a>) — see <a href='vis.immunr_hclust.html'>vis.immunr_hclust</a>;</p>
<p>- Repertoire incremental overlaps (from <a href='repOverlap.html'>repOverlap</a>) — see <a href='vis.immunr_top_overlap.html'>vis.immunr_top_overlap</a>;</p>
<p>- Public repertoire abundance (from pubRep) — vis <a href='vis.immunr_public_repertoire.html'>vis.immunr_public_repertoire</a>.</p>
<p>Gene usage:</p>
<p>- Gene usage statistics (from <a href='geneUsage.html'>geneUsage</a>) using bar plots, box plots, treemaps — see <a href='vis.immunr_gene_usage.html'>vis.immunr_gene_usage</a>;</p>
<p>- Gene usage distances (from <a href='geneUsageAnalysis.html'>geneUsageAnalysis</a>) using heatmaps, circos plots, polar area plots — see <a href='vis.immunr_ov_matrix.html'>vis.immunr_ov_matrix</a>;</p>
<p>- Gene usage clustering (from <a href='geneUsageAnalysis.html'>geneUsageAnalysis</a>) — see <a href='vis.immunr_hclust.html'>vis.immunr_hclust</a>.</p>
<p>Diversity estimation:</p>
<p>- Diversity estimations (from <a href='repDiversity.html'>repDiversity</a>) — see <a href='vis.immunr_chao1.html'>vis.immunr_chao1</a>.</p>
<p>Advanced analysis:</p>
<p>- Repertoire dynamics (from <a href='trackClonotypes.html'>trackClonotypes</a>) — see <a href='vis.immunr_dynamics.html'>vis.immunr_dynamics</a>;</p>
<p>- Sequence logo plots of amino acid distributions (from <a href='split_to_kmers.html'>kmer_profile</a>) — see <a href='vis_logo.html'>vis_logo</a>;</p>
<p>- Kmers distributions (from <a href='getKmers.html'>getKmers</a>) — see vis.immunr_kmers;</p>
<p>- Mutation networks (from mutationNetwork) — see <a href='vis.immunr_mutation_network.html'>vis.immunr_mutation_network</a>;</p>
<p>- CDR3 amino acid properties, e.g., biophysical (from <a href='cdrProp.html'>cdrProp</a>) — see <a href='vis.immunr_cdr_prop.html'>vis.immunr_cdr_prop</a>.</p>
<p>Additionaly, we provide a wrapper functions for visualisations of common data types:</p>
<p>- Any data frames or matrices using heatmaps — see <a href='vis_heatmap.html'>vis_heatmap</a> and <a href='vis_heatmap2.html'>vis_heatmap2</a>;</p>
<p>- Any data frames or matrices using circos plots — see <a href='vis_circos.html'>vis_circos</a>;</p>
<p>- Any data frames or matrices using polar area plots — see <a href='vis_radar.html'>vis_radar</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='fixVis.html'>fixVis</a> for precise manipulation of plots.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># Load the test data</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>immdata</span>)

<span class='co'># Compute and visualise overlaps</span>
<span class='no'>ov</span> <span class='kw'>=</span> <span class='fu'><a href='repOverlap.html'>repOverlap</a></span>(<span class='no'>immdata</span>$<span class='no'>data</span>)
<span class='fu'>vis</span>(<span class='no'>ov</span>)
<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by  ImmunoMind.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

